File: src/net/zomis/tttultimate/TTFactories.java
package net.zomis.tttultimate;

public class TTFactories {
	private static final TTMNKParameters mnkEmpty = new TTMNKParameters(0, 0, 0);
	
	private static final TicFactory lastFactory = new TicFactory() {
		@Override
		public TTBase construct(TTBase parent, int x, int y) {
			return new TTBase(parent, x, y, mnkEmpty, null);
		}
	};
	private static final TicFactory areaFactory = new TicFactory() {
		@Override
		public TTBase construct(TTBase parent, int x, int y) {
			return new TTBase(parent, x, y, parent.getMNKParameters(), lastFactory);
		}
	};
	
	public static class Factory implements TicFactory {
		private final TTMNKParameters	mnk;
		private final TicFactory	next;
		public Factory(TTMNKParameters mnk, TicFactory nextFactory) {
			this.mnk = mnk;
			this.next = nextFactory;
		}
		@Override
		public TTBase construct(TTBase parent, int x, int y) {
			return new TTBase(parent, x, y, mnk, next);
		}
	}
	
	public TTBase classicMNK(int width, int height, int consecutive) {
		return new TTBase(null, new TTMNKParameters(width, height, consecutive), lastFactory);
	}
	public TTBase classicMNK(int mnk) {
		return classicMNK(mnk, mnk, mnk);
	}
	public TTBase ultimate(int mnk) {
		return ultimate(mnk, mnk, mnk);
	}
	public TTBase ultimate(int width, int height, int consecutive) {
		return new TTBase(null, new TTMNKParameters(width, height, consecutive), areaFactory);
	}
	public TTBase ultimate() {
		return ultimate(3);
	}
	public TTBase othello(int size) {
		return new TTBase(null, new TTMNKParameters(size, size, size + 1), lastFactory);
	}
	

}

File: uttt-common/src/main/kotlin/net/zomis/tttultimate/TTFactories.kt
package net.zomis.tttultimate

class TTFactories {

    fun factory(mnk: TTMNKParameters, next: TicFactory): TicFactory {
        return {parent, x, y -> TTBase(parent, x, y, mnk, next) }
    }

    fun classicMNK(width: Int, height: Int, consecutive: Int): TTBase {
        return TTBase(null, TTMNKParameters(width, height, consecutive), lastFactory)
    }

    fun classicMNK(mnk: Int): TTBase {
        return classicMNK(mnk, mnk, mnk)
    }

    fun ultimate(mnk: Int = 3): TTBase {
        return ultimateMNK(mnk, mnk, mnk)
    }

    fun ultimateMNK(width: Int, height: Int, consecutive: Int): TTBase {
        return TTBase(null, TTMNKParameters(width, height, consecutive), areaFactory)
    }

    fun othello(size: Int): TTBase {
        return TTBase(null, TTMNKParameters(size, size, size + 1), lastFactory)
    }

    private val mnkEmpty = TTMNKParameters(0, 0, 0)

    private val lastFactory: TicFactory = {parent, x, y ->
        TTBase(parent, x, y, mnkEmpty, null)
    }

    private val areaFactory: TicFactory = {parent, x, y ->
        TTBase(parent, x, y, parent.mnkParameters, lastFactory)
    }

}
